File: data/derived_cot/rq3_prompting/results/translation/qwen/java_to_cpp/code_output/MetricsCalculator.cpp
#include <vector>
#include <iostream>

class MetricsCalculator {
private:
    int truePositives;
    int falsePositives;
    int falseNegatives;
    int trueNegatives;

public:
    MetricsCalculator() : truePositives(0), falsePositives(0), falseNegatives(0), trueNegatives(0) {}

    void update(const std::vector<int>& predictedLabels, const std::vector<int>& trueLabels, int length) {
        for (int i = 0; i < length; i++) {
            if (predictedLabels[i] == 1 && trueLabels[i] == 1) {
                truePositives++;
            } else if (predictedLabels[i] == 1 && trueLabels[i] == 0) {
                falsePositives++;
            } else if (predictedLabels[i] == 0 && trueLabels[i] == 1) {
                falseNegatives++;
            } else if (predictedLabels[i] == 0 && trueLabels[i] == 0) {
                trueNegatives++;
            }
        }
    }

    double precision(const std::vector<int>& predictedLabels, const std::vector<int>& trueLabels) {
        update(predictedLabels, trueLabels, predictedLabels.size());
        if (truePositives + falsePositives == 0) {
            return 0.0;
        }
        return static_cast<double>(truePositives) / (truePositives + falsePositives);
    }

    double recall(const std::vector<int>& predictedLabels, const std::vector<int>& trueLabels) {
        update(predictedLabels, trueLabels, predictedLabels.size());
        if (truePositives + falseNegatives == 0) {
            return 0.0;
        }
        return static_cast<double>(truePositives) / (truePositives + falseNegatives);
    }

    double f1Score(const std::vector<int>& predictedLabels, const std::vector<int>& trueLabels) {
        update(predictedLabels, trueLabels, predictedLabels.size());
        double precision = precision(predictedLabels, trueLabels);
        double recall = recall(predictedLabels, trueLabels);
        if (precision + recall == 0.0) {
            return 0.0;
        }
        return (2 * precision * recall) / (precision + recall);
    }

    double accuracy(const std::vector<int>& predictedLabels, const std::vector<int>& trueLabels) {
        update(predictedLabels, trueLabels, predictedLabels.size());
        int total = truePositives + trueNegatives + falsePositives + falseNegatives;
        if (total == 0) {
            return 0.0;
        }
        return static_cast<double>(truePositives + trueNegatives) / total;
    }

    static void main() {
        MetricsCalculator mc;
        std::vector<int> predictedLabels = {1, 1, 0, 0};
        std::vector<int> trueLabels = {1, 0, 0, 1};
        std::cout << mc.precision(predictedLabels, trueLabels) << std::endl;
        std::cout << mc.recall(predictedLabels, trueLabels) << std::endl;
        std::cout << mc.f1Score(predictedLabels, trueLabels) << std::endl;
        std::cout << mc.accuracy(predictedLabels, trueLabels) << std::endl;
    }
};

int main() {
    MetricsCalculator::main();
    return 0;
}

File: data/derived_cot/rq3_prompting/results/translation/qwen/java_to_py/code_output/MetricsCalculator.py
class MetricsCalculator:
    def __init__(self):
        self.true_positives = 0
        self.false_positives = 0
        self.false_negatives = 0
        self.true_negatives = 0

    def update(self, predicted_labels, true_labels):
        for i in range(len(predicted_labels)):
            if predicted_labels[i] == 1 and true_labels[i] == 1:
                self.true_positives += 1
            elif predicted_labels[i] == 1 and true_labels[i] == 0:
                self.false_positives += 1
            elif predicted_labels[i] == 0 and true_labels[i] == 1:
                self.false_negatives += 1
            elif predicted_labels[i] == 0 and true_labels[i] == 0:
                self.true_negatives += 1

    def precision(self, predicted_labels, true_labels):
        self.update(predicted_labels, true_labels)
        total_positive = self.true_positives + self.false_positives
        if total_positive == 0:
            return 0.0
        return float(self.true_positives) / total_positive

    def recall(self, predicted_labels, true_labels):
        self.update(predicted_labels, true_labels)
        total_positive = self.true_positives + self.false_negatives
        if total_positive == 0:
            return 0.0
        return float(self.true_positives) / total_positive

    def f1_score(self, predicted_labels, true_labels):
        self.update(predicted_labels, true_labels)
        precision_val = self.precision(predicted_labels, true_labels)
        recall_val = self.recall(predicted_labels, true_labels)
        if precision_val + recall_val == 0.0:
            return 0.0
        return (2 * precision_val * recall_val) / (precision_val + recall_val)

    def accuracy(self, predicted_labels, true_labels):
        self.update(predicted_labels, true_labels)
        total = self.true_positives + self.true_negatives + self.false_positives + self.false_negatives
        if total == 0:
            return 0.0
        return float(self.true_positives + self.true_negatives) / total


if __name__ == "__main__":
    mc = MetricsCalculator()
    predicted_labels = [1, 1, 0, 0]
    true_labels = [1, 0, 0, 1]
    print(mc.precision(predicted_labels, true_labels))
    print(mc.recall(predicted_labels, true_labels))
    print(mc.f1_score(predicted_labels, true_labels))
    print(mc.accuracy(predicted_labels, true_labels))